riscv-executor/src/lib.rs

```diff
@@ -645,59 +904,202 @@ struct Executor<'a, 'b, F: FieldElement> {
+    program_cols: HashMap<String, String>,
+    step: u32,
+    mode: ExecMode,
 }
 
 impl<'a, 'b, F: FieldElement> Executor<'a, 'b, F> {
+    fn init(&mut self) {
+        self.step = 4;
+        for i in 0..2 {
+            for (fixed, col) in &self.program_cols {
+                let val = Elem::Field(
+                    *self
+                        .get_fixed(fixed)
+                        .unwrap_or(&Vec::new())
+                        .get(i as usize)
+                        .unwrap_or(&F::zero()),
+                );
+                self.proc.set_col_idx(col, i as usize, val);
+            }
+            self.proc
+                .set_col_idx("main::_operation_id", i as usize, 2.into());
+            self.proc
+                .set_col_idx("main::pc_update", i as usize, (i + 1).into());
+        }
+    }
+
+    fn get_fixed(&self, name: &str) -> Option<&Vec<F>> {
+        self.fixed
+            .iter()
+            .find(|(n, _)| n == name)
+            // ROM is uniquely sized, which for now is all we looking at
+            .map(|(_, v)| v.get_uniquely_sized().expect("not uniquely sized!"))
+    }
+
+    fn sink_id(&self) -> u32 {
+        // sink_id is the length of the ROM. we find it by looking at the line instr__loop activates
+        for (i, &val) in self
+            .get_fixed("main__rom::p_instr__loop")
+            .unwrap()
+            .iter()
+            .enumerate()
+        {
+            if val.is_one() {
+                return i as u32;
+            }
+        }
+        panic!("could not find sink_id by looking at the p_instr__loop column");
+    }
+
+    /// read register value, updating the register memory machine
+    fn reg_read(&mut self, step_offset: u32, reg: u32, selector_idx: u32) -> Elem<F> {
+        let val = self.proc.get_reg_mem(reg);
+        self.proc
+            .regs_machine
+            .read(self.step + step_offset, reg, val, selector_idx);
+        val
+    }
+
+    /// write value to register, updating the register memory machine
+    fn reg_write(&mut self, step_offset: u32, reg: u32, val: Elem<F>, selector_idx: u32) {
+        if let ExecMode::Trace = self.mode {
+            self.proc
+                .regs_machine
+                .write(self.step + step_offset, reg, val, selector_idx);
+        }
+        self.proc.set_reg_mem(reg, val);
+    }
+
+    fn set_program_columns(&mut self, pc: u32) {
+        if let ExecMode::Trace = self.mode {
+            // set witness from the program definition
+            for (fixed, col) in &self.program_cols {
+                let val = Elem::Field(
+                    *self
+                        .get_fixed(fixed)
+                        .unwrap_or(&Vec::new())
+                        .get(pc as usize)
+                        .unwrap_or(&F::zero()),
+                );
+                self.proc.set_col(col, val);
+            }
+            // always 2 because RISCV only uses constrainted submachines
```

Comment:
Don't really get what this means

Reply:
we're only ever executing inside `main` function in our "virtual" asm machine. If we were to call functions in some other "virtual machine" (non block/constrained machines), we'd get different operation id's here afaiu... but I will remove the comment as it's kind of obscure and unlikely to ever be an issue